active-rfcs/0000-recursive-components.md

```diff
@@ -0,0 +1,159 @@
+
+The Single File per Component (SFC) pattern that Astro uses is simple, but inflexible. 
+In other frameworks like React (and presumable Vue and SolidJs) you can create multiple 
+components within a single file. 
+This allows you to can create both a function/component to render an `<Item />` and another 
+function/component to render `<ItemChildren />` have them reference each other, and then expose 
+either/both as exports. You can't do this with SFC. And, if you try to create these as 
+2 separate files, you get circular dependencies. The only way for SFC to allow for recursion 
+is by allowing a component access to reference it's own render function.
+
+Svelte, which also uses SFC, has already
+encountered and solved for this issue by exposing the [svelte:self](https://svelte.dev/docs#svelte_self)
+attribute as part of their API.
+
+
+# Detailed design
+
+The `Astro.self` property exposes the render function of the component.
+Defining a constant in the frontmatter and setting it equal to this allows
+you to name this component whatever you wish.
+
+```
+// src/components/RecursiveItem.astro
+---
+const {itemId, items} = Astro.props;
+const {childIds} = items[itemId];
+const MyRecursiveItem = Astro.self;
+---
+<li>
+    Item: {itemId}
+    {childIds.length ? (
+        <ol>
+            {childIds.map(childId => (
+                <MyRecursiveItem itemId={childId} items={items} />
+            ))}
+        </ol>
+    ) : ""}
+</>
+```
+
+# Alternatives
+
+The other design pattern considered was to mimic the Svelte syntax, using
+`<astro:self />`
+
+This follows an already established precedent (for those familiar with Svelte),
+but there's nothing else in Astro yet that leverages `astro:` prefixes as
+components.
+
+By contrast, using `Astro.self` is more explicit,
+requiring a bit more work to tap into an advanced feature, which was marginally
+preferred by some in the discussion group. Also providing incremental value is
+the ability for developers to name the component how they wish (although
+this may also be a source of some minor confusion).
+
+Most importantly, it builds upon existing Astro conventions. We're already
+referencing things like `Astro.props` and `Astro.request.params` in the frontmatter.
+This is just one more `Astro` property that becomes available there.
+
+# Adoption strategy
+
+- No backwards compatibility concerns
+- No potential for breaking changes
+- Worth noting that recursive rendering can be already achieved with any number of other
+platforms: React, Vue, Svelte, etc., so we can approach this at our leisure with
+the short term answer to this being: "Do that in your own platform for now".
+- Need to update Documentation
+
+# Unresolved questions
+
+Svelte has some [additional restrictions](https://svelte.dev/docs#svelte_self)
+about how/where theirs can be used. It must be within a conditional or an "each" loop.
+They are attempting to protect against infinite loops. Should we build in similar protections?
```

Comment:
The answer to this question might push us to resolve the specific syntax for this feature.

From the Svelte docs:

> It cannot appear at the top level of your markup; it must be inside an if or each block or passed to a component's slot to prevent an infinite loop.

One potential problem with `Astro.self` is that we're unable to statically detect usage unless it is literally authored as `<Astro.self />` because reassignment would confuse the compiler.

On the other hand, `astro:self` would always be statically analyzable.

Alternatively, we could exclusively document the feature using `Astro.self` and warn when encountering any literal `<Astro.self />` usage. That would likely cover 80% of the uses.